File: src/logic/pieces/knight.py
from .piece import Piece

class Knight(Piece):
    def letter(self):
        return "n"

    def legal_moves(self, board: "Board") -> list["Move"]:
        ret = []
        for dx, dy in [
            (+2, +1), (+1, +2), # north east
            (+2, -1), (+1, -2), # south east
            (-2, -1), (-1, -2), # south west
            (-2, +1), (-1, +2), # north west
        ]:
            move = self._move_for_position(board, self.pos.x + dx, self.pos.y + dy)
            if move is not None:
                ret.append(move)
        return ret

File: cpp/src/model/pieces/knight.cpp
#include "../board/board.hpp"
#include "../utils/coords.hpp"
#include "../utils/move.hpp"
#include "piece.hpp"

#include <vector>

std::vector<Move> knight_moves(const Board& b, const Coords xy) {
    std::vector<Move> ret;
    std::vector<std::pair<int, int>> moves = {
        {+2, +1},
        {+1, +2}, // north east
        {+2, -1},
        {+1, -2}, // south east
        {-2, -1},
        {-1, -2}, // south west
        {-2, +1},
        {-1, +2} // north west
    };

    for (const auto& [dx, dy] : moves) {
        std::optional<Move> move =
            move_for_position(b, xy, Coords{xy.x + dx, xy.y + dy});
        if (move.has_value())
            ret.push_back(move.value());
    }
    return ret;
}
